lib/model/forum/ForumDradenMeldingModel.class.php

```diff
@@ -0,0 +1,192 @@
+ */
+class ForumDradenMeldingModel extends CachedPersistenceModel {
+
+	const ORM = ForumDraadMelding::class;
+
+	protected function maakForumDraadMelding($draad_id, $uid, $niveau) {
+		$melding = new ForumDraadMelding();
+		$melding->draad_id = $draad_id;
+		$melding->uid = $uid;
+		$melding->niveau = $niveau;
+		$this->create($melding);
+		return $melding;
+	}
+
+	public function getAltijdMeldingVoorDraad(ForumDraad $draad) {
+		return $this->prefetch('draad_id = ? AND niveau = ?', array($draad->draad_id, ForumDraadMeldingNiveau::ALTIJD));
+	}
+
+	public function getVoorkeursNiveauVoorLid(ForumDraad $draad, $uid = null) {
+		if ($uid === null) $uid = LoginModel::getUid();
+
+		/** @var ForumDraadMelding $voorkeur */
+		$voorkeur = $this->retrieveByPrimaryKey(array($draad->draad_id, $uid));
+		if ($voorkeur) {
+			return $voorkeur->niveau;
+		} else {
+			return false;
+		}
+	}
+
+	public function getNiveauVoorLid(ForumDraad $draad, $uid = null) {
+		if ($uid === null) $uid = LoginModel::getUid();
+
+		/** @var ForumDraadMelding $voorkeur */
+		$voorkeur = $this->retrieveByPrimaryKey(array($draad->draad_id, $uid));
+		if ($voorkeur) {
+			return $voorkeur->niveau;
+		} else {
+			$wilMeldingBijVermelding = LidInstellingenModel::getInstellingVoorLid('forum', 'meldingStandaard', $uid);
+			return $wilMeldingBijVermelding === 'ja' ? ForumDraadMeldingNiveau::VERMELDING : ForumDraadMeldingNiveau::NOOIT;
+		}
+	}
+
+	public function setNiveauVoorLid(ForumDraad $draad, $niveau) {
+		$uid = LoginModel::getUid();
+		/** @var ForumDraadMelding $voorkeur */
+		$voorkeur = $this->retrieveByPrimaryKey(array($draad->draad_id, $uid));
+		if ($voorkeur) {
+			$voorkeur->niveau = $niveau;
+			$this->update($voorkeur);
+		} else {
+			$this->maakForumDraadMelding($draad->draad_id, $uid, $niveau);
+		}
+	}
+
+	public function stopAlleMeldingenVoorLid($uid) {
+		foreach ($this->find('uid = ?', array($uid)) as $volgen) {
+			$this->delete($volgen);
+		}
+	}
+
+	public function stopMeldingenVoorIedereen($draad) {
+		foreach ($this->find('draad_id = ?', array($draad->draad_id)) as $volgen) {
+			$this->delete($volgen);
+		}
+	}
+
+	public function stuurMeldingen(ForumPost $post) {
+		$this->stuurMeldingenNaarVolgers($post);
+		$this->stuurMeldingenNaarGenoemden($post);
+	}
+
+	/**
+	 * Verzendt mail
+	 *
+	 * @param Profiel $ontvanger
+	 * @param Profiel $auteur
+	 * @param ForumPost $post
+	 * @param ForumDraad $draad
+	 * @param string $bericht
+	 */
+	private function stuurMelding($ontvanger, $auteur, $post, $draad, $bericht) {
+		$ontvangerNaam = $ontvanger->getNaam('civitas');
+		$values = array(
+			'NAAM' => $ontvangerNaam,
+			'AUTEUR' => $auteur->getNaam('civitas'),
+			'POSTLINK' => $post->getLink(true),
+			'TITEL' => $draad->titel,
+			'TEKST' => str_replace('\r\n', "\n", $post->tekst),
+		);
+
+		// Stel huidig UID in op ontvanger om te voorkomen dat ontvanger privé of andere persoonlijke info te zien krijgt
+		$curUID = $_SESSION['_uid'];
```

Comment:
Misschien is het beter om deze regels in loginmodel oid te zetten

Reply:
Bijvoorbeeld een functie om naar een gebruiker te switchen en een functie om weer terug te gaan?

Reply:
Ja, dan staat de `$_SESSION['_uid']` naast de andere plekken waar ie wordt gebruikt

Reply:
`overrideUid` `resetUid` ofzo

Reply:
Ook vrij gevaarlijke code denk ik: Als bijvoorbeeld $mail->send een error geeft (bijvoorbeeld door een fout in BB-code), dan blijft $_SESSION['_uid'] op de verkeerde gebruiker staan (denk ik). Effectief betekent dat dat je van gebruiker bent gewisseld.

Reply:
Scherp, een try-finally zou hier wel op z'n plek zijn dan

Reply:
Tijdens het ontwikkelen heb ik dit een aantal keer meegemaakt. Dit wordt enigszins afgevangen door de validatie waarin gecheckt wordt of het account in de huidige sessie dezelfde is als degene die uit het UID komt. Effectief logt hij het lid dus uit als dit fout gaat.

Het alternatief is via de mail class een uid doorgeven naar de parser en daar (net als in LidInstellingen) een overschrijfbaar uid maken, maar dan leek dit me een efficiëntere oplossing.

Heb jij nog andere ideeën?

Reply:
Ah, een try-finally zou kunnen werken, even proberen

Reply:
Werkt goed 👍